File: app-macos/Sources/WDashboardCore/Pomodoro.swift
// Pomodoro timer pure logic (docs/sdd.md §11, ADR-012).
//
// `pomodoroReduce` (state machine) and `pomodoroView` (state -> display data) are
// pure functions locked by the shared test vectors (docs/test-vectors/pomodoro*/),
// mirroring app-linux/src/pomodoro.rs. The 1s tick, menu-bar icon, notifications
// and sound live in the app layer (AppState / AppDelegate), not here.

/// After a `*Ended` phase has been unacknowledged for this long, `pomodoroReduce`
/// returns to `.idle` on its own — the user is assumed to have walked away
/// (docs/sdd.md §11.3 rule 6).
public let pomodoroAutoStopOvertimeSecs: Int64 = 1800

public enum PomodoroPhase: String, Equatable, Sendable, Codable {
    case idle = "Idle"
    case focus = "Focus"
    case brk = "Break"
    case focusEnded = "FocusEnded"
    case breakEnded = "BreakEnded"
}

public enum PomodoroEvent: String, Equatable, Sendable, Codable {
    case startFocus = "StartFocus"
    case startBreak = "StartBreak"
    case stop = "Stop"
    case tick = "Tick"
}

/// The single in-memory pomodoro state. Never persisted (ADR-012).
public struct PomodoroState: Equatable, Sendable {
    public var phase: PomodoroPhase
    /// Unix seconds the current `.focus`/`.brk` segment started; `nil` in `.idle`.
    /// Kept through `*Ended` so overtime can be measured.
    public var phaseStartedAt: Int64?
    /// Target focus length; the app writes this from config only while `.idle`, so
    /// a running segment's target is frozen (docs/sdd.md §11.3).
    public var focusSecs: Int64
    public var breakSecs: Int64

    public init(phase: PomodoroPhase, phaseStartedAt: Int64?, focusSecs: Int64, breakSecs: Int64) {
        self.phase = phase
        self.phaseStartedAt = phaseStartedAt
        self.focusSecs = focusSecs
        self.breakSecs = breakSecs
    }

    /// A fresh idle state with the given segment lengths.
    public static func idle(focusSecs: Int64, breakSecs: Int64) -> PomodoroState {
        PomodoroState(phase: .idle, phaseStartedAt: nil, focusSecs: focusSecs, breakSecs: breakSecs)
    }
}

/// Display data derived from a `PomodoroState` at `now` (docs/sdd.md §11.2).
public struct PomodoroView: Equatable, Sendable {
    public var phase: PomodoroPhase
    public var remainingSecs: Int64
    public var elapsedSecs: Int64
    public var overtimeSecs: Int64
    public var progress: Double
    public var alerting: Bool

    public init(
        phase: PomodoroPhase, remainingSecs: Int64, elapsedSecs: Int64,
        overtimeSecs: Int64, progress: Double, alerting: Bool
    ) {
        self.phase = phase
        self.remainingSecs = remainingSecs
        self.elapsedSecs = elapsedSecs
        self.overtimeSecs = overtimeSecs
        self.progress = progress
        self.alerting = alerting
    }
}

/// The frozen target length of the phase's segment: focus side vs break side.
private func phaseDuration(_ state: PomodoroState) -> Int64 {
    switch state.phase {
    case .focus, .focusEnded: return state.focusSecs
    case .brk, .breakEnded: return state.breakSecs
    case .idle: return 0
    }
}

private func isAlerting(_ phase: PomodoroPhase) -> Bool {
    phase == .focusEnded || phase == .breakEnded
}

/// State machine (docs/sdd.md §11.3). Pure; reads only `state`, never config.
public func pomodoroReduce(_ state: PomodoroState, _ event: PomodoroEvent, now: Int64) -> PomodoroState {
    var next = state
    switch event {
    case .startFocus:
        next.phase = .focus
        next.phaseStartedAt = now
    case .startBreak:
        next.phase = .brk
        next.phaseStartedAt = now
    case .stop:
        next.phase = .idle
        next.phaseStartedAt = nil
    case .tick:
        guard let started = state.phaseStartedAt else { return state }
        let raw = now - started
        switch state.phase {
        case .focus where raw >= state.focusSecs:
            next.phase = .focusEnded
        case .brk where raw >= state.breakSecs:
            next.phase = .breakEnded
        case .focusEnded, .breakEnded:
            guard raw - phaseDuration(state) >= pomodoroAutoStopOvertimeSecs else { return state }
            next.phase = .idle
            next.phaseStartedAt = nil
        default:
            return state
        }
    }
    return next
}

/// Display data derived from `state` at `now` (docs/sdd.md §11.2). Pure.
public func pomodoroView(_ state: PomodoroState, now: Int64) -> PomodoroView {
    let alerting = isAlerting(state.phase)

    guard let started = state.phaseStartedAt else {
        // Only reachable in .idle.
        return PomodoroView(
            phase: state.phase, remainingSecs: 0, elapsedSecs: 0,
            overtimeSecs: 0, progress: 0.0, alerting: alerting)
    }

    let raw = now - started
    let duration = phaseDuration(state)

    switch state.phase {
    case .idle:
        return PomodoroView(
            phase: state.phase, remainingSecs: 0, elapsedSecs: 0,
            overtimeSecs: 0, progress: 0.0, alerting: alerting)
    case .focus, .brk:
        return PomodoroView(
            phase: state.phase,
            remainingSecs: max(0, duration - raw),
            elapsedSecs: raw,
            overtimeSecs: 0,
            progress: min(1.0, max(0.0, Double(raw) / Double(duration))),
            alerting: alerting)
    case .focusEnded, .breakEnded:
        return PomodoroView(
            phase: state.phase,
            remainingSecs: 0,
            elapsedSecs: duration,
            overtimeSecs: max(0, raw - duration),
            progress: 1.0,
            alerting: alerting)
    }
}

File: app-linux/src/pomodoro.rs
//! Pomodoro timer pure logic (docs/sdd.md §11, ADR-012).
//!
//! This module is the deterministic core: `pomodoro_reduce` (state machine) and
//! `pomodoro_view` (state -> display data). Both are pure functions locked by the
//! shared test vectors (`docs/test-vectors/pomodoro*/`). The 1s tick, tray icon,
//! notifications and sound live in the UI layer (`main.rs`), never here.

/// After a `*Ended` phase has been unacknowledged for this long, `pomodoro_reduce`
/// returns to `Idle` on its own — the user is assumed to have walked away
/// (docs/sdd.md §11.3 rule 6).
pub const AUTO_STOP_OVERTIME_SECS: i64 = 1800;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PomodoroPhase {
    Idle,
    Focus,
    Break,
    /// Focus time is up, waiting for the user to acknowledge (icon flashes).
    FocusEnded,
    /// Break time is up, waiting for the user to acknowledge (icon flashes).
    BreakEnded,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PomodoroEvent {
    StartFocus,
    StartBreak,
    Stop,
    Tick,
}

/// The single in-memory pomodoro state. Never persisted (ADR-012).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct PomodoroState {
    pub phase: PomodoroPhase,
    /// Unix seconds the current `Focus`/`Break` segment started; `None` in `Idle`.
    /// Kept through `*Ended` so overtime can be measured.
    pub phase_started_at: Option<i64>,
    /// Target focus length; the UI writes this from config only while `Idle`, so a
    /// running segment's target is frozen (docs/sdd.md §11.3).
    pub focus_secs: i64,
    pub break_secs: i64,
}

impl PomodoroState {
    /// A fresh idle state with the given segment lengths.
    pub fn idle(focus_secs: i64, break_secs: i64) -> Self {
        PomodoroState {
            phase: PomodoroPhase::Idle,
            phase_started_at: None,
            focus_secs,
            break_secs,
        }
    }
}

/// Display data derived from `state` at `now` (docs/sdd.md §11.2).
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct PomodoroView {
    pub phase: PomodoroPhase,
    pub remaining_secs: i64,
    pub elapsed_secs: i64,
    pub overtime_secs: i64,
    pub progress: f64,
    pub alerting: bool,
}

/// Display data derived from `state` at `now` (docs/sdd.md §11.2). Pure.
pub fn pomodoro_view(state: &PomodoroState, now: i64) -> PomodoroView {
    let alerting = matches!(
        state.phase,
        PomodoroPhase::FocusEnded | PomodoroPhase::BreakEnded
    );

    let Some(started) = state.phase_started_at else {
        // Only reachable in Idle.
        return PomodoroView {
            phase: state.phase,
            remaining_secs: 0,
            elapsed_secs: 0,
            overtime_secs: 0,
            progress: 0.0,
            alerting,
        };
    };

    let raw = now - started;
    let duration = phase_duration(state);

    match state.phase {
        PomodoroPhase::Idle => PomodoroView {
            phase: state.phase,
            remaining_secs: 0,
            elapsed_secs: 0,
            overtime_secs: 0,
            progress: 0.0,
            alerting,
        },
        PomodoroPhase::Focus | PomodoroPhase::Break => PomodoroView {
            phase: state.phase,
            remaining_secs: (duration - raw).max(0),
            elapsed_secs: raw,
            overtime_secs: 0,
            progress: (raw as f64 / duration as f64).clamp(0.0, 1.0),
            alerting,
        },
        PomodoroPhase::FocusEnded | PomodoroPhase::BreakEnded => PomodoroView {
            phase: state.phase,
            remaining_secs: 0,
            elapsed_secs: duration,
            overtime_secs: (raw - duration).max(0),
            progress: 1.0,
            alerting,
        },
    }
}

/// The frozen target length of the phase's segment: focus side vs break side.
fn phase_duration(state: &PomodoroState) -> i64 {
    match state.phase {
        PomodoroPhase::Focus | PomodoroPhase::FocusEnded => state.focus_secs,
        PomodoroPhase::Break | PomodoroPhase::BreakEnded => state.break_secs,
        PomodoroPhase::Idle => 0,
    }
}

/// State machine (docs/sdd.md §11.3). Pure; reads only `state`, never config.
pub fn pomodoro_reduce(state: PomodoroState, event: PomodoroEvent, now: i64) -> PomodoroState {
    match event {
        PomodoroEvent::StartFocus => PomodoroState {
            phase: PomodoroPhase::Focus,
            phase_started_at: Some(now),
            ..state
        },
        PomodoroEvent::StartBreak => PomodoroState {
            phase: PomodoroPhase::Break,
            phase_started_at: Some(now),
            ..state
        },
        PomodoroEvent::Stop => PomodoroState {
            phase: PomodoroPhase::Idle,
            phase_started_at: None,
            ..state
        },
        PomodoroEvent::Tick => {
            let Some(started) = state.phase_started_at else {
                return state;
            };
            let raw = now - started;
            match state.phase {
                PomodoroPhase::Focus if raw >= state.focus_secs => PomodoroState {
                    phase: PomodoroPhase::FocusEnded,
                    ..state
                },
                PomodoroPhase::Break if raw >= state.break_secs => PomodoroState {
                    phase: PomodoroPhase::BreakEnded,
                    ..state
                },
                PomodoroPhase::FocusEnded | PomodoroPhase::BreakEnded
                    if raw - phase_duration(&state) >= AUTO_STOP_OVERTIME_SECS =>
                {
                    PomodoroState {
                        phase: PomodoroPhase::Idle,
                        phase_started_at: None,
                        ..state
                    }
                }
                _ => state,
            }
        }
    }
}
